fix HDF5Reader build with MSVC

germasch opened the pull request:
Sadly, variable-sized stack arrays are only a feature of C99, but not C++11,
so this patch uses std::vector instead.

This fixes #1485, the second issue there, at least hopefully so -- @ax3l, could you test this one, too, as I don't think the adios2 CI covers this case. (I changed the place identified by the error message from your report, and one more place with the same issue, but there may be more).

@chuckatkins, I haven't checked, but my guess is that the MSVC CI doesn't build ADIOS2 with HDF5, otherwise I would have expected to see this problem earlier. So that may be something for the future todo list.

germasch commented:
@pnorbert, changed the base to release here, too. Same thing, though.

germasch commented:
Though, for this PR, it'd be good to hear back from @ax3l as to whether it really fixes the hdf5 engine build troubles he was seeing with MSVC, as I couldn't do that testing (and I think the adios2 CI doesn't, either.)

I'm pretty sure that it fixes the one particular error he posted, and I found one more similar occasion, but there may be issues beyond that.

ax3l commented:
Ok, there is a little bit more in the HDF5 files that we need to address: https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/25636465
```
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(192): error C2065: 'uint': undeclared identifier
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(192): error C2146: syntax error: missing ';' before identifier 'totalAdiosSteps'
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(192): error C2065: 'totalAdiosSteps': undeclared identifier
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(196): error C2065: 'totalAdiosSteps': undeclared identifier
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(199): error C2065: 'totalAdiosSteps': undeclared identifier
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(230): warning C4244: '=': conversion from 'hsize_t' to 'unsigned int', possible loss of data
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(257): warning C4018: '<': signed/unsigned mismatch
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(281): warning C4244: '=': conversion from 'ssize_t' to 'herr_t', possible loss of data
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(298): error C2131: expression did not evaluate to a constant
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(298): note: failure was caused by non-constant arguments or reference to a non-constant symbol
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(298): note: see usage of 'heritage'
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(300): error C2131: expression did not evaluate to a constant
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(298): note: failure was caused by non-constant arguments or reference to a non-constant symbol
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(298): note: see usage of 'heritage'
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(381): warning C4244: '=': conversion from 'ssize_t' to 'herr_t', possible loss of data
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(596): warning C4018: '>=': signed/unsigned mismatch
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(1038): warning C4267: 'initializing': conversion from 'size_t' to 'int', possible loss of data
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(1130): warning C4018: '<': signed/unsigned mismatch
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(1251): error C2131: expression did not evaluate to a constant
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(1251): note: failure was caused by non-constant arguments or reference to a non-constant symbol
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(1251): note: see usage of 'MAX_ATTR_NAME_SIZE'
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(1309): error C2131: expression did not evaluate to a constant
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(1309): note: failure was caused by non-constant arguments or reference to a non-constant symbol
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(1309): note: see usage of 'MAX_ATTR_NAME_SIZE'
```

ax3l commented:
Urgh, sorry I am of so little help. Here we go with two more:
```
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(936): error C2131: expression did not evaluate to a constant

%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(938): error C2664: 'adios2::core::Attribute<T[arraySize]> &adios2::core::IO::DefineAttribute<T[arraySize]>(const std::string &,const signed char (&)[arraySize],const std::string &,const std::string)': cannot convert argument 3 from 'hsize_t' to 'const std::string &'
        with
        [
            T=int8_t
        ]
%SRC_DIR%\source\adios2\toolkit\interop\hdf5\HDF5Common.cpp(938): note: Reason: cannot convert from 'hsize_t' to 'const std::string'

```

germasch commented:
Rebased onto latest release, and fixed more variable length arrays -- should have gotten all this time via `-Werror=vla`.

I should note that I'm doing these patches kinda in a rush, and sometimes not exactly sober, in between things like putting on a tuxedo and top hat to walk through 35 centigrade weather (maybe some Germans here understand...) Anyway, it certainly wouldn't be a bad idea to give these patches a good second look.

germasch commented:
> Please don't forget to `#include <vector>` in all of these files :-)

I'm pretty sure I did do all those includes this time ;)

I just force-pushed to re-run he AppVeyor CI, which I think hit just a random hiccup.

williamfgc commented:
superseeded by #1587
